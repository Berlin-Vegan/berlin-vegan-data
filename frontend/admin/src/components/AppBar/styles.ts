import { makeStyles } from '@material-ui/core/styles';
import { drawerWidth } from '../PageContainer/styles';

export const useStyles = makeStyles((theme) => ({
  toolbar: {
    paddingRight: 24, // keep right padding when drawer closed
  },
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
    transition: theme.transitions.create(['width', 'margin'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
  },
  appBarShift: {
    marginLeft: drawerWidth,
    width: `calc(100% - ${drawerWidth}px)`,
    transition: theme.transitions.create(['width', 'margin'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
  title: {
    flexGrow: 1,
  },
  menuButton: {
    marginRight: 36,
  },
  menuButtonHidden: {
    display: 'none',
  },
  logo: {
    marginRight: 20,
    width: 50,
    height: 50,
  },
  logout: {
    margin: theme.spacing(1),
  },
}));
