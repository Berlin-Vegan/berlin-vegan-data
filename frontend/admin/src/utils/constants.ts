export const GASTRO: string = 'GASTRO';
export const SHOPPING: string = 'SHOPPING';

export type LocationType = typeof GASTRO | typeof SHOPPING;

const API = '/api/';
export const V1 = `${API}v1/`;
export const V2 = `${API}v2/`;
