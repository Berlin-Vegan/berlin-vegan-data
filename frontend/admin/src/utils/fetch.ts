import { getCSRFToken } from './cookie';
import { UserDispatch, TYPE_USER_LOGOUT } from '../providers/UserProvider';

export const authorizedFetch = async (
  userDispatch: UserDispatch,
  url: string,
  method: string = 'GET',
  body: object | Array<unknown> | null = null
): Promise<Response> => {
  const response = await fetch(url, {
    method,
    body: body ? JSON.stringify(body) : body,
    headers: {
      'Content-Type': 'application/json',
      'X-CSRFToken': getCSRFToken(),
    },
  });

  if (response.status === 403) {
    userDispatch({ type: TYPE_USER_LOGOUT });
  }

  return response;
};

export const fetchGastroList = async (
  userDispatch: UserDispatch,
  filter: string
): Promise<Response> =>
  authorizedFetch(userDispatch, `/api/v1/gastros/?${filter}`);
