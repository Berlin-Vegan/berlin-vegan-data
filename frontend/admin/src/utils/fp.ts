import {
  addIndex,
  compose,
  curry,
  defaultTo,
  join,
  map,
  nth,
  pipe,
  splitAt,
  toLower,
  toUpper,
} from 'ramda';

export const nthOr = curry((defaultVal, list) =>
  compose(defaultTo(defaultVal), nth)(list),
);

export const nthOrEmpty = nthOr('');

export const mapIndexed = addIndex(map);

export const capitalizeWord = pipe(
  toLower,
  splitAt(1),
  (splitted) => [toUpper(nthOrEmpty(0)(splitted)), nthOrEmpty(1)(splitted)],
  join(''),
);
