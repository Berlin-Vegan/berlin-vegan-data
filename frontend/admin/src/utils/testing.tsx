import renderer from 'react-test-renderer';
import React, { ComponentType } from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';

export const standardComponentTest = (C: ComponentType<any>, props = {}) => {
  renderWithTestRenderer(C, props);
  it('renders without crashing', () => renders(C, props));
};

export const standardComponentTestMockFetch = (
  C: ComponentType<any>,
  props = {},
  fakeResponse: any,
) => {
  renderWithTestRenderer(C, props);
  it('renders without crashing and mock fetch', async () => {
    jest.spyOn(window, 'fetch').mockImplementation((): Promise<any> => {
      const fetchResponse = {
        json: () => Promise.resolve(fakeResponse),
      };
      return Promise.resolve(fetchResponse);
    });
    await act(async () => renders(C, props));
  });
};

export const renders = (C: ComponentType, props: object) => {
  const div = document.createElement('div');
  ReactDOM.render(<C {...props} />, div);
};

const getName = (C: ComponentType) => C.displayName || C.name || 'A Component';

const renderWithTestRenderer = (C: ComponentType, props: object) => {
  it(`render <${getName(C)} />`, () => {
    const tree = renderer.create(<C {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
};
