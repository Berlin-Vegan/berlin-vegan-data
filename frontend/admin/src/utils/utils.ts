import { pipe, toLower } from 'ramda';
import { LocationType, V1, V2 } from './constants';

//
export const buildReviewListUrl: string = `${V1}review/`;
export const buildReviewDetailUrl = (id: number): string =>
  `${buildReviewListUrl}${id}/`;

// API V2 urls
export const buildListUrl = (type: LocationType): string =>
  pipe(toLower, (lowerType: string) => `${V2}${lowerType}/`)(type);
export const buildDetailUrl = (type: LocationType, id: string): string =>
  pipe(toLower, buildListUrl, (baseUrl: string) => `${baseUrl}${id}/`)(type);
// FE urls
export const buildFEDetailUrl = (type: LocationType, id: string): string =>
  pipe(toLower, (lowerType: string) => `/${lowerType}/${id}/edit`)(type);
