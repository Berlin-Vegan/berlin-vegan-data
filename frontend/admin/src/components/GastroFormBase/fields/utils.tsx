import React from 'react';
import MenuItem from '@material-ui/core/MenuItem';
import { map } from 'ramda';
import { nthOr } from '../../../utils/fp';

const menuItem = (value: number | string, label: string) => (
  <MenuItem value={value} key={value}>
    {label}
  </MenuItem>
);

export const menuItems = (fieldOptions: Array<[number | string, string]>) =>
  map((item) => menuItem(nthOr('', 0)(item), nthOr('', 1)(item)))(fieldOptions);

export const withLeadingZero = (value: number): string =>
  value < 10 ? `0${value}` : `${value}`;
