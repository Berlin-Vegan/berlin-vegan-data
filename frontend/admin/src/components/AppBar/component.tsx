import React, { FunctionComponent, useContext } from 'react';
import clsx from 'clsx';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import ExitToAppIcon from '@material-ui/icons/ExitToApp';
import { Button } from '@material-ui/core';

import { pathOr } from 'ramda';
import { useStyles } from './styles';
import Logo from '../../berlin-vegan-logo.png';
import { authorizedFetch } from '../../utils/fetch';
import {
  AuthContext,
  UserDispatch,
  TYPE_USER_LOGOUT,
} from '../../providers/UserProvider';

type AppBarComponentProps = {
  open: boolean;
  handleDrawerOpen: () => void;
};

const logoutUser = async (userDispatch: UserDispatch): Promise<void> => {
  const response = await authorizedFetch(
    userDispatch,
    '/api/v1/accounts/logout/',
    'POST'
  );
  if (response.ok) {
    userDispatch({ type: TYPE_USER_LOGOUT });
  }
};

const AppBarComponent: FunctionComponent<AppBarComponentProps> = ({
  open,
  handleDrawerOpen,
}) => {
  const classes = useStyles();
  const { state, dispatch: userDispatch } = useContext(AuthContext);
  const username = pathOr('â€“', ['userData', 'username'], state);

  return (
    <AppBar
      position="absolute"
      className={clsx(classes.appBar, open && classes.appBarShift)}
    >
      <Toolbar className={classes.toolbar}>
        <IconButton
          edge="start"
          color="inherit"
          aria-label="open drawer"
          onClick={handleDrawerOpen}
          className={clsx(classes.menuButton, open && classes.menuButtonHidden)}
        >
          <MenuIcon />
        </IconButton>
        <img src={Logo} alt="Berlin-Vegan Logo" className={classes.logo} />
        <Typography
          component="h1"
          variant="h6"
          color="inherit"
          noWrap
          className={classes.title}
        >
          Berlin-Vegan Data
        </Typography>
        <Typography>{username}</Typography>
        <Button
          variant="contained"
          color="secondary"
          className={classes.logout}
          startIcon={<ExitToAppIcon />}
          onClick={() => logoutUser(userDispatch)}
        >
          Logout
        </Button>
      </Toolbar>
    </AppBar>
  );
};

export default AppBarComponent;
