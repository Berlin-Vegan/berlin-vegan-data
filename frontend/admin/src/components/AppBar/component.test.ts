import AppBar from './component';
import { standardComponentTest } from '../../utils/testing';

standardComponentTest(AppBar);
