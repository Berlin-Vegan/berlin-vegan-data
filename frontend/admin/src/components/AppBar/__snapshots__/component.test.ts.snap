// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render <AppBarComponent /> 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionAbsolute MuiAppBar-colorPrimary makeStyles-appBar-2 MuiPaper-elevation4"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular makeStyles-toolbar-1 MuiToolbar-gutters"
  >
    <button
      aria-label="open drawer"
      className="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-5 MuiIconButton-colorInherit MuiIconButton-edgeStart"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
      </span>
    </button>
    <img
      alt="Berlin-Vegan Logo"
      className="makeStyles-logo-7"
      src="berlin-vegan-logo.png"
    />
    <h1
      className="MuiTypography-root makeStyles-title-4 MuiTypography-h6 MuiTypography-colorInherit MuiTypography-noWrap"
    >
      Berlin-Vegan Data
    </h1>
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      â€“
    </p>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-logout-8 MuiButton-containedSecondary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        Logout
      </span>
    </button>
  </div>
</header>
`;
